File: server/apollo/src/core/budget.ts
import {Account} from './account';
import {Balance} from './balance';
import {Month} from './month';
import {Transfer} from './transfer';

export class Budget {
  // Period over which the budget is defined.
  months: Month[] = [];
  // Account name to account map.
  readonly accounts = new Map<string, Account>();
  // Account name -> month -> balance map.
  readonly balances = new Map<string, Map<string, Balance>>();
  // Account name -> month -> transfers map.
  readonly transfers = new Map<string, Map<string, Set<Transfer>>>();

  setPeriod(start: Month, end: Month) {
    this.months = Array.from(Month.generate(start, end));
  }

  setAccount(account: Account) {
    this.accounts.set(account.name, account);
  }

  setBalance(accountName: string, month: string, balance: Balance) {
    let balances = this.balances.get(accountName);
    if (!balances) {
      balances = new Map<string, Balance>();
      this.balances.set(accountName, balances);
    }
    balances.set(month, balance);
  }
}

File: server/kt/src/main/kotlin/com/cherba29/tally/core/Budget.kt
package com.cherba29.tally.core

/**
 * Data representing all accounts, their transactions and totals per month.
 */
data class Budget(
  /** Period over which the budget is defined. */
  val months: MonthRange,
  /** Hierarchical structure of accounts and summaries. */
  val tree: Group,
  /** Maps leaf tree node to corresponding account. */
  val leafToAccount: Map<Group.Leaf, Account>,
  // Account id to account map.
  val accounts: Map<NodeId, Account>,
  // Account id -> month -> balance map.
  val balances: Map<NodeId, Map<Month, Balance>>,
  // Account id -> month -> transfers map.
  val transfers: Map<NodeId, Map<Month, List<Transfer>>>
)